### ISSUE TITLE
单卡改多卡

### ISSUE THREAD

**Issue by JJ-Guo1996** [yzyouzhang/AIR-ASVspoof#32]
请问将代码改为多卡时，需要将ocsoftmax-loss 包装到DP里面吗？

**Comment by yzyouzhang** [comment]
我认为需要，如果出现问题，请告诉我

**Comment by JJ-Guo1996** [comment]
试了DP和DDP,包装loss和不包装loss都没有达到加速的效果呢。


Junjun Guo
***@***.***
发自 网易邮箱大师



---- 回复的原邮件 ----
发件人 You ***@***.***>日期 2022年08月14日 12:11收件人 ***@***.***>抄送至 ***@***.******@***.***>主题 Re: [yzyouzhang/AIR-ASVspoof] 单卡改多卡 (Issue #32)

我认为需要，如果出现问题，请告诉我

**Comment by JJ-Guo1996** [comment]
> 

试了DP和DDP,包装loss和不包装loss都没有达到加速的效果呢。

**Comment by yzyouzhang** [comment]
可以分享一下你的代码和测试结果吗？我觉得DP和DDP是可以加速的

**Comment by JJ-Guo1996** [comment]
> 可以分享一下你的代码和测试结果吗？我觉得DP和DDP是可以加速的

现在可以加速了，但是eer变大了2个点。现在eer = 4左右。
 lfcc_model = ResNet(3, args.enc_dim, resnet_type='18', nclasses=2)
 lfcc_model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(lfcc_model)
 lfcc_model = DDP(lfcc_model.cuda(args.local_rank),device_ids=[args.local_rank], find_unused_parameters=True)

    if args.add_loss == "ocsoftmax":
        ocsoftmax = OCSoftmax(args.enc_dim, r_real=args.r_real, r_fake=args.r_fake, alpha=args.alpha) 
        ocsoftmax = DDP(ocsoftmax.cuda(args.local_rank),device_ids=[args.local_rank], find_unused_parameters=True)
        ocsoftmax.train()
        ocsoftmax_optimzer = torch.optim.SGD(ocsoftmax.parameters(), lr=args.lr)

**Comment by yzyouzhang** [comment]
感谢分享，我觉得可能要重新写个class把ResNet和OC-Softmax放在一起，这样才能保证每一张卡上的这两个权重是同步并且对应更新的
